src/server/mod.rs

```diff
@@ -1 +1,279 @@
-pub mod net;
+extern crate chan;
+extern crate rand;
+
+use ::bincode;
+use ::bincode::serde::{deserialize_from, DeserializeError};
+use ::model;
+use ::model::types::{Board, Player, PlayerType, SubField};
+use ::net::{self, types};
+use ::net::types::{MessageType};
+use ::util;
+use std::net::{Shutdown, TcpListener, TcpStream};
+use std::sync::atomic::AtomicBool;
+use std::sync::Arc;
+use term_painter::ToStyle;
+use term_painter::Color::*;
+
+#[derive(Debug)]
+enum CurrentPlayer {
+    Host,
+    Client,
+}
+
+/// Initialize and prepare game
+pub fn init(name: String, size: u8, board: Vec<SubField>) {
+    let listener = TcpListener::bind((types::LOCALHOST, types::DEFAULT_PORT)).unwrap();
+
+    // accept one incoming connection
+    let (client_conn, _) = listener.accept().unwrap();
+    let mut client_stream = client_conn.try_clone().unwrap();
+
+    // welcome client
+    net::send(
+        &mut client_stream,
+        MessageType::Welcome(
+            "Welcome stranger, let me sink your ships!".to_string(),
+            name.clone())
+    );
+
+    // add CTRL+C system hook, so that connection partner is informed about disconnect
+    let client_stream_clone = client_stream.try_clone().unwrap();
+    let running = Arc::new(AtomicBool::new(true));
+    let _ = running.clone();
+    ::ctrlc::set_handler(move || {
+        net::send(&mut client_stream_clone.try_clone().unwrap(), MessageType::Quit);
+        client_stream_clone.shutdown(Shutdown::Both).expect("shutdown call failed");
+    }).expect("Error setting Ctrl+C handler");
+
+    // wait for client to send his name
+    let recv: Result<types::MessageType, DeserializeError> =
+        deserialize_from(&mut client_stream, bincode::SizeLimit::Infinite);
+    let client_name = match recv {
+        Ok(received) => {
+            match received {
+                MessageType::Login(name) => {
+                    name
+                },
+                MessageType::Quit => {
+                    println!("Client closed connection.");
+                    return
+                },
+                _ => {
+                    // unexpected packet
+
+                    "".to_string()
+                },
+            }
+        },
+        Err(_) => {
+            println!("ERROR hit me");
+            "".to_string()
+        },
+    };
+
+    // create players
+    let host = Player {
+        own_board: board.clone(),
+        op_board: Board::init(),
+        player_type: PlayerType::Human,
+        name: name,
+        capacity: Board::targets(&board),
+    };
+
+    let client = Player {
+        own_board: Board::init(),
+        op_board: Board::init(),
```

Comment:
Quite right, but how to accomplish that if we want the server to perform all the logic (Hit, Miss, Won?). And we want to send just relevant changes between the player and not the whole board.
Correct me if I'm mistaken, but then I need to save the Client data on the Server and in the Client instance so that the Client can have access to its own data. Hmm...